////////////////////////////////////////////////////////////////////////////////
// Copyright (c) 2014-2019, Lawrence Livermore National Security, LLC.
// Produced at the Lawrence Livermore National Laboratory.
// Written by the LBANN Research Team (B. Van Essen, et al.) listed in
// the CONTRIBUTORS file. <lbann-dev@llnl.gov>
//
// LLNL-CODE-697807.
// All rights reserved.
//
// This file is part of LBANN: Livermore Big Artificial Neural Network
// Toolkit. For details, see http://software.llnl.gov/LBANN or
// https://github.com/LLNL/LBANN.
//
// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
// may not use this file except in compliance with the License.  You may
// obtain a copy of the License at:
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
// implied. See the License for the specific language governing
// permissions and limitations under the license.
////////////////////////////////////////////////////////////////////////////////

syntax = "proto3";

package lbann_data;

import "callbacks.proto";

message LbannPB {
  DataReader data_reader = 1;
  Model model = 2;
  Optimizer optimizer = 3;
  MotifDefinitions motif_definitions = 4;
  DataSetMetaData data_set_metadata = 5;
}

//========================================================================
// DataReaders
//========================================================================
message DataReader {
  int64 max_par_io_size = 1;
  repeated Reader reader = 2;
  bool requires_data_set_metadata = 3;
}

message Reader {
  string name = 1; //mnist, nci, nci_regression, numpy, imagenet, synthetic, merge_samples
  string role = 3; //train, validation, test
  bool shuffle = 4;
  string data_filedir = 5;
  string data_local_filedir = 50; //to support data_store
  string data_filename = 6;
  string label_filename = 7;
  string index_list = 8;
  double validation_percent = 9;
  int64 absolute_sample_count = 11;
  int64 first_n = 200;
  double percent_of_data_to_use = 12;

  //for GAN model
  bool gan_labelling = 201;
  int32 gan_label_value = 202;

  int32 num_labels = 99; //for imagenet and synthetic
  int64 num_samples = 100; //only for synthetic
  string synth_dimensions = 101; //only for synthetic
  string synth_response_dimensions = 115; //only for synthetic
  //csv attributes
  string separator = 102;
  int32 skip_cols = 103;
  int32 skip_rows = 104;
  bool has_header = 105;
  int32 label_col = 106;
  int32 response_col = 107;
  bool disable_labels = 108;
  bool disable_responses = 109;
  string format = 110; // numpy, csv
  string data_file_pattern = 111;
  int64 num_neighbors = 112; // pilot2_molecular_reader
  int64 max_neighborhood = 113; // pilot2_molecular_reader
  int32 num_image_srcs = 114; // data_reader_multi_images
  float scaling_factor_int16 = 116; // for numpy_npz_reader with int16 data

  int32 max_files_to_load = 1000;

  //------------- start of only for partitioned data sets ------------------
  bool is_partitioned = 300;
  double partition_overlap = 301;
  int32 partition_mode = 302;
       // 1 - share a portion of your data with two neighbors;
       // 2 - there's a set of overlap indices that are common to all models
  //------------- end of only for partitioned data sets ------------------

  //------------- start of only for index lists ------------------
  bool index_list_per_trainer = 400;
  bool index_list_per_model   = 401;
  //------------- end of only for index lists ------------------

  PythonDataReader python = 501;

  repeated Transform transforms = 600;  // Ordered list of transforms to apply.
}

message PythonDataReader {
  string module = 1;                // Python module
  string module_dir = 2;            // Directory containing Python module
  string sample_function = 3;       // Function that gets data sample
  string num_samples_function = 4;  // Function that gets number of data samples
  string sample_dims_function = 5;  // Function that gets dimensions of data sample
}

// Preprocessing transforms.
message Transform {
  // Transforms that apply to LBANN data.
  // Normalize channel-wise with mean and standard deviation.
  message Normalize {
    string means = 1;
    string stddevs = 2;
  }
  // Normalize each sample to have mean 0, standard deviation 1.
  message SampleNormalize {}
  // Scale by a constant.
  message Scale {
    float scale = 1;
  }

  // Transforms that apply to images.
  // Adjust the brightness of an image.
  message AdjustBrightness {
    float factor = 1;
  }
  // Adjust the contrast of an image.
  message AdjustContrast {
    float factor = 1;
  }
  // Adjust the saturation of an image.
  message AdjustSaturation {
    float factor = 1;
  }
  // Crop of size height x width from the center.
  message CenterCrop {
    uint64 height = 1;
    uint64 width = 2;
  }
  // Convert to color.
  message Colorize {}
  // Randomly jitter brightness/contrast/saturation.
  message ColorJitter {
    float min_brightness_factor = 1;
    float max_brightness_factor = 2;
    float min_contrast_factor = 3;
    float max_contrast_factor = 4;
    float min_saturation_factor = 5;
    float max_saturation_factor = 6;
  }
  // Apply cutout augmentation.
  message Cutout {
    uint64 num_holes = 1;
    uint64 length = 2;
  }
  // Convert to grayscale.
  message Grayscale {}
  // Horizontal flip with probability p.
  message HorizontalFlip {
    float p = 1;
  }
  // Fused Normalize + ToLBANNLayout.
  message NormalizeToLBANNLayout {
    string means = 1;
    string stddevs = 2;
  }
  // Apply a random affine transform.
  message RandomAffine {
    float rotate_min = 1;
    float rotate_max = 2;
    float translate_h = 3;
    float translate_w = 4;
    float scale_min = 5;
    float scale_max = 6;
    float shear_min = 7;
    float shear_max = 8;
  }
  // Crop of size height x width from a random location.
  message RandomCrop {
    uint64 height = 1;
    uint64 width = 2;
  }
  // Random crop with scale and aspect ratio augmentation.
  message RandomResizedCrop {
    uint64 height = 1;
    uint64 width = 2;
    float scale_min = 3;
    float scale_max = 4;
    float ar_min = 5;
    float ar_max = 6;
  }
  // Resize to height x width, then randomly crop to crop_height x crop_width.
  message RandomResizedCropWithFixedAspectRatio {
    uint64 height = 1;
    uint64 width = 2;
    uint64 crop_height = 3;
    uint64 crop_width = 4;
  }
  // Resize to height x width.
  message Resize {
    uint64 height = 1;
    uint64 width = 2;
  }
  // Resize to height x width then crop to crop_height x crop_width at the center.
  message ResizedCenterCrop {
    uint64 height = 1;
    uint64 width = 2;
    uint64 crop_height = 3;
    uint64 crop_width = 4;
  }
  // Convert from an image to LBANN data.
  message ToLBANNLayout { }
  // Vertical flip with probability p.
  message VerticalFlip {
    float p = 1;
  }

  oneof a_transform {
    // On LBANN data:
    Normalize normalize = 1;
    SampleNormalize sample_normalize = 2;
    Scale scale = 3;

    // On images:
    CenterCrop center_crop = 100;
    Colorize colorize = 101;
    Grayscale grayscale = 102;
    HorizontalFlip horizontal_flip = 103;
    NormalizeToLBANNLayout normalize_to_lbann_layout = 104;
    RandomAffine random_affine = 105;
    RandomCrop random_crop = 106;
    RandomResizedCrop random_resized_crop = 107;
    RandomResizedCropWithFixedAspectRatio random_resized_crop_with_fixed_aspect_ratio = 108;
    Resize resize = 109;
    ResizedCenterCrop resized_center_crop = 110;
    ToLBANNLayout to_lbann_layout = 111;
    VerticalFlip vertical_flip = 112;
    AdjustBrightness adjust_brightness = 113;
    AdjustContrast adjust_contrast = 114;
    AdjustSaturation adjust_saturation = 115;
    ColorJitter color_jitter = 116;
    Cutout cutout = 117;
  }
}

//========================================================================
// Metadata for a Data set
//========================================================================
message DataSetMetaData {
  message Schema {
    string scalar_prefix = 1;
    string image_prefix = 2;
    string input_prefix = 3;

    uint64 image_height = 11;
    uint64 image_width = 12;
    uint64 image_num_channels = 13;

    //------------------ start of only for jag_conduit -----------------------
    bool split_jag_image_channels = 89;
    repeated string jag_image_keys = 90;
    repeated string jag_scalar_keys = 91;
    repeated string jag_input_keys = 92;
    message JagKeyPrefixFilter {
      string key_prefix = 1;
      uint32 min_len = 2;
    }
    repeated string jag_scalar_filters = 93;
    repeated JagKeyPrefixFilter jag_scalar_prefix_filters = 94;
    repeated string jag_input_filters = 95;
    repeated JagKeyPrefixFilter jag_input_prefix_filters = 96;

    enum JAG_Data {
      Undefined  = 0;
      JAG_Image  = 1;
      JAG_Scalar = 2;
      JAG_Input  = 3;
    }
    message JAGDataSlice {
      repeated JAG_Data pieces = 1;
    }
    repeated JAGDataSlice independent = 97;
    repeated JAGDataSlice dependent = 98;
    //------------------  end of only for jag_conduit  -----------------------
  }

  message Normalization {
    //------------------ start of only for jag_conduit -----------------------
    message JagLinearNormalizationParams {
      double scale = 1;
      double bias = 2;
    }

    repeated JagLinearNormalizationParams jag_image_normalization_params = 86;
    repeated JagLinearNormalizationParams jag_scalar_normalization_params = 87;
    repeated JagLinearNormalizationParams jag_input_normalization_params = 88;

    //------------------  end of only for jag_conduit  -----------------------
  }
    Schema schema = 1;
  Normalization normalization = 2;
}

//========================================================================
// Model
//========================================================================

message Model {
  string type = 1;
  string name = 3;
  ObjectiveFunction objective_function = 2;
  repeated Metric metric = 5;
  string data_layout = 6;
  bool shareable_training_data_reader = 42; // Can the data reader be shared across multiple models( e.g. GAN)
  bool shareable_testing_data_reader = 43; // Can the data reader be shared across multiple models (e.g. GAN)
  bool shareable_validation_data_reader = 44; // Can the data reader be shared across multiple models (e.g. GAN)

  int64 mini_batch_size = 12;
  int64 num_epochs = 4;
  int64 super_steps = 121; //multiple steps/epochs currently use in GAN
  int64 num_batches = 122; //multiple batches/sub epoch
  int64 block_size = 50;
  int64 procs_per_trainer = 51;
  int64 num_gpus = 53; //has no effect
  int64 evaluation_frequency = 54;
  int64 num_parallel_readers = 100;
  bool  serialize_io = 101;

  bool disable_cuda = 8;

  repeated Layer layer = 10;

  repeated Weights weights = 11;

  repeated Callback callback = 20;

  int64 random_seed = 30;
  // If true, models will have their model rank mixed into their random seed.
  bool random_init_models_differently = 31;

}

//========================================================================
// Objective function
//========================================================================

message ObjectiveFunction {
  repeated LayerTerm layer_term = 1;
  repeated L2WeightRegularization l2_weight_regularization = 2;
}

message LayerTerm {
  double scale_factor = 1;
  string layer = 2;
}

message L2WeightRegularization {
  double scale_factor = 1;
  string weights = 2;   // If empty, L2 regularization is applied to all weights
}

//========================================================================
// Metrics
//========================================================================

message Metric {
  LayerMetric layer_metric = 11;
}

message LayerMetric {
  string layer = 1;
  string name = 2;
  string unit = 3;
}

//========================================================================
// Optimizers
//========================================================================
message Optimizer {
  oneof optimizer_type {
    AdaGrad adagrad = 1;
    Adam adam = 2;
    HypergradientAdam hypergradient_adam = 3;
    RMSprop rmsprop = 4;
    SGD sgd = 5;
  }
}

message AdaGrad {
  double learn_rate = 1;
  double eps = 2;           // Suggested: 1e-8
}

message Adam {
  double learn_rate = 1;
  double beta1 = 6;         // Suggested: 0.9
  double beta2 = 7;         // Suggested: 0.99
  double eps = 8;           // Suggested: 1e-8
}

message HypergradientAdam {
  double init_learning_rate = 1;
  double hyper_learning_rate = 2;   // Suggested: 1e-7
  double beta1 = 6;                 // Suggested: 0.9
  double beta2 = 7;                 // Suggested: 0.99
  double eps = 8;                   // Suggested: 1e-8
}

message RMSprop {
  double learn_rate = 1;
  double decay_rate = 2;
  double eps = 3;           // Suggested: 1e-8
}

message SGD {
  double learn_rate = 1;
  double momentum = 2;      // Set to zero for vanilla SGD
  bool nesterov = 4;
}

//========================================================================
// Weights
//========================================================================

message Weights {

  string name = 1;
  Optimizer optimizer = 2;

  ConstantInitializer constant_initializer = 20;
  ValueInitializer value_initializer = 21;
  UniformInitializer uniform_initializer = 22;
  NormalInitializer normal_initializer = 23;
  GlorotNormalInitializer glorot_normal_initializer = 24;
  GlorotUniformInitializer glorot_uniform_initializer = 25;
  HeNormalInitializer he_normal_initializer = 26;
  HeUniformInitializer he_uniform_initializer = 27;
  LeCunNormalInitializer lecun_normal_initializer = 28;
  LeCunUniformInitializer lecun_uniform_initializer = 29;

}

// Weight initializers
message ConstantInitializer {
  double value = 1;
}
message ValueInitializer {
  string values = 1;
}
message UniformInitializer {
  double min = 1;
  double max = 2;
}
message NormalInitializer {
  double mean = 1;
  double standard_deviation = 2;
}
message GlorotNormalInitializer {}
message GlorotUniformInitializer {}
message HeNormalInitializer {}
message HeUniformInitializer {}
message LeCunNormalInitializer {}
message LeCunUniformInitializer {}

//note: I'd like to put this enum inside of Layer, but if I do the enum values
//      become, e.g, Layer_Imcomm_EXCLUDE, which is just ugly
enum Imcomm {
  DEFAULT = 0; //add Layer to Imcomm callback if all_learning_layers = true in
               //the CallbackImComm
  EXCLUDE = 1; //*do not* add Layer to Imcomm callback if all_learning_layers = true in
               //the CallbackImComm
  INCLUDE = 2;  //add Layer to Imcomm callback regardless of whether all_learning_layers
                //in the CallbackImComm is set to true or false
}

// Weight data for exporting
message WeightsShape {
  repeated int64 dim = 1 [packed = true];
}
message WeightsData {
  WeightsShape shape = 5;
  string name = 1;
  int64 height = 2;
  int64 width = 3;
  //@todo assume float above, add other datatype
  repeated float data = 4 [packed=true];

  Imcomm imcomm = 55;
}

//========================================================================
// MotifDefinitions
//========================================================================

message MotifDefinitions {
  repeated Motif motif = 1;
}

message Motif {
  string name = 1;
  repeated Layer layer = 2;
}

//========================================================================
// Layers
//========================================================================

message Layer {
   string name = 50;
   string parents = 151;
   string children = 152;
   string data_layout = 52;
   string device_allocation = 55;
   string weights = 54;
   bool num_neurons_from_data_reader = 53;
   bool freeze = 5;
   string hint_layer = 56;

   repeated WeightsData weights_data = 153;
   string top = 154;
   string bottom = 155;
   string type = 156;

   // a Layer should contain exactly one of the following
   // (this may or may not be properly checked for in proto_common.cpp)
   //
   // @todo: this should be done better using oneof:
   //   oneof a_layer {
   //       Reshape reshape = 306
   //       Pooling pooling = 12;
   //       ...
   //   }
   //
   //

   // motif layer
   MotifLayer motif_layer = 4;

   // Input layers
   Input input = 2;

   // Transform layers
   Reshape reshape = 306;
   Pooling pooling = 12;
   Concatenation concatenation = 300;
   Slice slice = 301;
   Split split = 302;
   Sum sum = 303;
   WeightedSum weighted_sum = 323;
   Unpooling unpooling = 304;
   Hadamard hadamard = 308;
   Constant constant = 309;
   Reduction reduction = 310;
   Evaluation evaluation = 311;
   Gaussian gaussian = 312;
   Bernoulli bernoulli = 313;
   Uniform uniform = 314;
   Crop crop = 316;
   CategoricalRandom categorical_random = 317;
   DiscreteRandom discrete_random = 318;
   Dummy dummy = 319;
   StopGradient stop_gradient = 320;
   InTopK in_top_k = 324;
   Sort sort = 325;
   WeightsLayer weights_layer = 326;
   Tessellate tessellate = 327;

   // Learning layers
   FullyConnected fully_connected = 11;
   Convolution convolution = 13;
   Deconvolution deconvolution = 305;
   Embedding embedding = 328;

   // Loss layers
   CrossEntropy cross_entropy = 60;
   MeanSquaredError mean_squared_error = 61;
   MeanAbsoluteError mean_absolute_error = 62;
   CategoricalAccuracy categorical_accuracy = 63;
   TopKCategoricalAccuracy top_k_categorical_accuracy = 64;
   L2Norm2 l2_norm2 = 65;
   L1Norm l1_norm = 66;
   BinaryCrossEntropy binary_cross_entropy = 67;
   SigmoidBinaryCrossEntropy sigmoid_binary_cross_entropy = 68;
   BooleanAccuracy boolean_accuracy = 69;
   BooleanFalseNegative boolean_false_negative = 70;
   BooleanFalsePositive boolean_false_positive = 71;

   // Math layers
   LogicalNot logical_not = 401;
   Abs abs = 402;
   Negative negative = 403;
   Sign sign = 404;
   Round round = 405;
   Ceil ceil = 406;
   Floor floor = 407;
   Reciprocal reciprocal = 408;
   Square square = 409;
   Sqrt sqrt = 410;
   Rsqrt rsqrt = 411;
   SafeReciprocal safe_reciprocal = 412;
   Exp exp = 413;
   Expm1 expm1 = 414;
   Log log = 415;
   Log1p log1p = 416;
   Cos cos = 417;
   Sin sin = 418;
   Tan tan = 419;
   Acos acos = 420;
   Asin asin = 421;
   Atan atan = 422;
   Cosh cosh = 423;
   Sinh sinh = 424;
   Tanh tanh = 425;
   Acosh acosh = 426;
   Asinh asinh = 427;
   Atanh atanh = 428;
   Add add = 450;
   Subtract subtract = 451;
   Multiply multiply = 452;
   Divide divide = 453;
   Mod mod = 454;
   Pow pow = 455;
   SafeDivide safe_divide = 456;
   SquaredDifference squared_difference = 457;
   Max max = 458;
   Min min = 459;
   Equal equal = 460;
   NotEqual not_equal = 461;
   Less less = 462;
   LessEqual less_equal = 463;
   Greater greater = 464;
   GreaterEqual greater_equal = 465;
   LogicalAnd logical_and = 466;
   LogicalOr logical_or = 467;
   LogicalXor logical_xor = 468;
   Clamp clamp = 469;

   // Regularization layers
   BatchNormalization batch_normalization = 19;
   LocalResponseNormalization local_response_normalization = 20;
   Dropout dropout = 21;
   SeluDropout selu_dropout = 229;

   // Activation layers
   Elu elu = 200;
   Identity identity = 201;
   LeakyRelu leaky_relu = 202;
   LogSigmoid log_sigmoid = 203;
   LogSoftmax log_softmax = 204;
   Relu relu = 205;
   Selu selu = 206;
   Sigmoid sigmoid = 207;
   Softmax softmax = 208;
   Softplus softplus = 209;
   Softsign softsign = 210;

   // Image layers
   BilinearResize bilinear_resize = 500;

   // Miscellaneous layers
   Covariance covariance = 600;
   Variance variance = 601;
   ChannelwiseMean channelwise_mean = 602;
   MiniBatchIndex mini_batch_index = 603;
   MiniBatchSize mini_batch_size = 604;

}
///////////////////////
// MotifLayer //
///////////////////////
message MotifLayer {
  string motif_id = 1;
  repeated string variable = 2;
}

///////////////////////
// Math layers       //
///////////////////////
message LogicalNot {}
message Abs {}
message Negative {}
message Sign {}
message Round {}
message Ceil {}
message Floor {}
message Reciprocal {}
message Square {}
message Sqrt {}
message Rsqrt {}
message SafeReciprocal {}
message Exp {}
message Expm1 {}
message Log {}
message Log1p {}
message Cos {}
message Sin {}
message Tan {}
message Acos {}
message Asin {}
message Atan {}
message Cosh {}
message Sinh {}
message Tanh {}
message Acosh {}
message Asinh {}
message Atanh {}
message Add {}
message Subtract {}
message Multiply {}
message Divide {}
message Mod {}
message Pow {}
message SafeDivide {}
message SquaredDifference {}
message Max {}
message Min {}
message Equal {}
message NotEqual {}
message Less {}
message LessEqual {}
message Greater {}
message GreaterEqual {}
message LogicalAnd {}
message LogicalOr {}
message LogicalXor {}
message Clamp {
  double min = 1;
  double max = 2;
}

///////////////////////
// Activation layers //
///////////////////////
message Elu {
  double alpha = 1; //default: 1.0; should be >= 0
}
message Identity {}
message LeakyRelu {
  double negative_slope = 1; //default: 0.01
}
message LogSigmoid {}
message LogSoftmax {}
message Relu {}
message Selu {}
message Sigmoid {}
message Softmax {}
message Softplus {}
message Softsign {}

///////////////////////
// Loss layers //
///////////////////////
message CrossEntropy {}
message MeanSquaredError {}
message MeanAbsoluteError {}
message CategoricalAccuracy {}
message TopKCategoricalAccuracy {
  int64 k = 1;
}
message L2Norm2 {}
message L1Norm {}
message BinaryCrossEntropy {}
message SigmoidBinaryCrossEntropy {}
message BooleanAccuracy {}
message BooleanFalseNegative {}
message BooleanFalsePositive {}

///////////////////////////
// Regularization layers //
///////////////////////////
message BatchNormalization {
  double decay = 1;          //default: 0.9
  double scale_init = 2;     //default: 1.0
  double bias_init = 3;      //default: 0.0
  double epsilon = 4;        //default: 1e-5
  string stats_aggregation = 5; // default: local
}

message SeluDropout {
  double keep_prob = 2; //default: 0.95
  double alpha = 3;     //default: 1.6732632423543772848170429916717
  double scale = 4;     //default: 1.0507009873554804934193349852946
}

message LocalResponseNormalization {
  int64 window_width = 4;
  double lrn_alpha = 5;
  double lrn_beta = 6;
  double lrn_k = 7;
}

message Dropout {
  double keep_prob = 2;  //default: 0.5
}

//////////////////
// Input layers //
//////////////////
message Input {
  bool data_set_per_model = 1;  // Default: false
  string io_buffer = 2;         // Options: "partitioned" (default)
  string target_mode = 3;       // Options: "classification" (default), "regression", "reconstruction", "N/A"
}

//////////////////////
// Transform layers //
//////////////////////
message Reshape {
  int64 num_dims = 1; //DEPRECATED
  string dims = 2; //should be space-separated list of ints, e.g, "2 6 7"
}

message Pooling {
  int64 num_dims = 1;

  bool has_vectors = 2;

  //these are used if has_vectors = true
  string pool_dims = 4; //should be space-separated list, e.g, "2 2 3"
  string pool_pads = 5; //should be space-separated list, e.g, "2 2 3"
  string pool_strides = 6; //should be space-separated list, e.g, "2 2 3"

  //these are used if has_vectors = false
  int64 pool_dims_i = 10;
  int64 pool_pads_i = 11;
  int64 pool_strides_i = 12;

  //pool_mode should be one of: max, average, average_no_pad
  //see: lbann/include/lbann/lbann_base.hpp
  string pool_mode = 7;
}

message Unpooling {
  int64 num_dims = 1;
  string pooling_layer = 13; //should be name of the pooling layer
}


message Concatenation {
  int64 axis = 1;
}

message Slice {
  int64 axis = 1;
  string slice_points = 2; //should be space-separated list of ints, e.g, "2 6 7"
  //the following is for jag_conduit_hdf5;
  string get_slice_points_from_reader = 4;
  bool get_slice_points_from_reader_bool = 5;
}

message Split {
}

message Sum {
}

message WeightedSum {
  string scaling_factors = 1; //should be a space-separated list of doubles, e.g. "1.0 2.0 -1.0"
}

message Hadamard {
}

message Constant {
  double value=1;
  string num_neurons=2;
}

message Reduction {
  string mode=1; //"sum" or "average"
}

message Evaluation {
}

message Gaussian {
  double mean = 1;
  double stdev = 2;
  string neuron_dims = 3;
}

message Bernoulli {
  double prob = 1;
  string neuron_dims = 2;
}

message Uniform {
  double min = 1;
  double max = 2;
  string neuron_dims = 3;
}


message Crop {
  string dims = 3;
}

message CategoricalRandom {
}

message DiscreteRandom {
  string values = 1;
  string dims = 2;
}

message Dummy {
}

message StopGradient {
}

message InTopK {
  int64 k = 1;
}

message Sort {
  bool descending = 1;
}

message WeightsLayer {
  string dims = 1;
}

message Tessellate {
  string dims = 1;
}

/////////////////////
// Learning layers //
/////////////////////
message FullyConnected {
  int64 num_neurons = 1;
  string weight_initialization = 2;    //DEPRECATED
  bool has_bias = 3;                   //default: true
  double bias_initial_value = 4;       //default: 0
  double l2_regularization_factor = 5; //default: 0
  double group_lasso_regularization_factor = 6; //default: 0
  bool transpose = 7;
  bool num_neurons_is_num_labels = 8;

  bool get_input_dimension_from_reader = 9;
  bool get_image_and_scalar_dimension_from_reader = 10;
  bool get_image_dimension_from_reader = 11;
  bool get_scalar_dimension_from_reader = 12;
  repeated uint32 get_num_neurons_of_slice_from_reader = 13;
  string get_slice_points_from_reader = 14;
}

message Convolution {
  int64 num_dims = 1;
  int64 num_output_channels = 4;
  int64 num_groups = 3;

  bool has_vectors = 2;

  // these are used if has_vector = true
  string conv_dims = 5; //should be space-separated list, e.g, "2 2 3"
  string conv_pads = 6;  //should be space-separated list, e.g, "2 2 3"
  string conv_strides = 7; //should be space-separated list, e.g, "2 2 3"
  string conv_dilations = 8;  //should be space-separated list, e.g. "2 3 3"

  // these are used if has_vector = false
  int64 conv_dims_i = 50;
  int64 conv_pads_i = 60;
  int64 conv_strides_i = 70;
  int64 conv_dilations_i = 80;

  string weight_initialization = 9;     //DEPRECATED
  bool has_bias = 10;                   //default: true
  double bias_initial_value = 11;       //default: 0
  double l2_regularization_factor = 12; //default: 0
}

message Deconvolution {
  int64 num_dims = 1;
  int64 num_output_channels = 4;
  int64 num_groups = 3;

  bool has_vectors = 2;

  // these are used if has_vector = true
  string conv_dims = 5; //should be space-separated list, e.g, "2 2 3"
  string conv_pads = 6;  //should be space-separated list, e.g, "2 2 3"
  string conv_strides = 7; //should be space-separated list, e.g, "2 2 3"
  string conv_dilations = 8;  //should be space-separated list, e.g. "2 3 3"

  // these are used if has_vector = false
  int64 conv_dims_i = 50;
  int64 conv_pads_i = 60;
  int64 conv_strides_i = 70;
  int64 conv_dilations_i = 80;

  string weight_initialization = 9;     //DEPRECATED
  bool has_bias = 10;                   //default: true
  double bias_initial_value = 11;       //default: 0
  double l2_regularization_factor = 12; //default: 0
}

message Embedding {
  int64 dictionary_size = 1;
  int64 embedding_size = 2;
}

//////////////////
// Image layers //
//////////////////
message BilinearResize {
  int64 height = 1;
  int64 width = 2;
}

//////////////////////////
// Miscellaneous layers //
//////////////////////////
message Covariance {
  bool biased = 1; //Whether to use a biased covariance estimate
}
message Variance {
  bool biased = 1; //Whether to use a biased variance estimate
}
message ChannelwiseMean {}
message MiniBatchIndex {}
message MiniBatchSize {}
